### ISSUE TITLE
Unit checking doesn't occur, and unit prefix handling leads to unexpected behavior

### ISSUE THREAD

**Issue by span17** [opencor/opencor#1317]
Currently units don't appear to be checked, for example the following code is considered valid:

        var T: newton {init:1};
        var Y: meter;
        Y = T;

Furthermore, prefixes don't appear to be handled, which can cause unexpected behavior in, for example, case statements:

    def unit millisecond as
        unit second {pref: milli};
    enddef;

        var F: rad_per_sec3;
        F = sel
            case (t > 0{second}) and (t < 1{millisecond}):
                1.0{rad_per_sec3};
            case (t > 1{second}) and (t < 2{second}):
                -1.0{rad_per_sec3};
            otherwise:
                0.0{rad_per_sec3};
        endsel;

If you look at a plot of F in the simulation panel, the first domain (0s-1ms) has the same extent as the second domain (1s-2s) instead of 1000x smaller.

Presumably this also affects variable definitions that use prefixes in some of the units involved.

**Comment by agarny** [comment]
Thanks for your report. Just as a general rule, it would help a lot if you were to tell us which version of OpenCOR you are using, as well as on which operating system. Also, and whenever possible, it would help if you could provide us with some CellML file(s) that illustrate your problem. Finally, to provide us with *exact* steps to reproduce your problem is critical.

Now, this being all said, I am going to assume that you are using the 2017-05-07 snapshot. Personally, I primarily use Windows and macOS, but hopefully the OS won't make any difference in this particular case.

> Currently units don't appear to be checked, for example the following code is considered valid:
> 
>         var T: newton {init:1};
>         var Y: meter;
>         Y = T;

Ok, I have just created a very simple model using the `CellML Text` view. The model reads:

```
def model test as
    def comp component as
        var T: newton {init: 1};
        var Y: meter;

        Y = T;
    enddef;
enddef;
```

Now, if I try to validate the model (by pressing `Ctrl+T`), I get the following message:

![screen shot 2017-05-24 at 09 51 11](https://cloud.githubusercontent.com/assets/602265/26392335/8844495a-4066-11e7-933e-dd5c0ab46232.png)

This is the expected result (when using the `CellML Text` view). Validation in this view is only done against the CellML Text format, not against CellML itself (as mentioned in the message above). So, to do the latter, you (currently) need to switch to the `Raw CellML` view. From there, if you validate again, you will get:

![screen shot 2017-05-24 at 09 53 05](https://cloud.githubusercontent.com/assets/602265/26392424/d4b4299a-4066-11e7-8e01-8ff50747bd26.png)

So, again, everything is considered to be fine from a CellML viewpoint. However, you will see a warning that reads:

```
[Warning] [8:18] MathML equals element has inconsistent units between the sides.
```

So, yes, CellML only warns about unit inconsistencies (although, as a general rule, I would always recommend addressing unit inconsistencies).

> Furthermore, prefixes don't appear to be handled, which can cause unexpected behavior in, for example, case statements:
> 
>     def unit millisecond as
>         unit second {pref: milli};
>     enddef;
> 
>         var F: rad_per_sec3;
>         F = sel
>             case (t > 0{second}) and (t < 1{millisecond}):
>                 1.0{rad_per_sec3};
>             case (t > 1{second}) and (t < 2{second}):
>                 -1.0{rad_per_sec3};
>             otherwise:
>                 0.0{rad_per_sec3};
>         endsel;
> 
> If you look at a plot of F in the simulation panel, the first domain (0s-1ms) has the same extent as the second domain (1s-2s) instead of 1000x smaller.
> 
> Presumably this also affects variable definitions that use prefixes in some of the units involved.

This time, I have created a small CellML file using the `CellML Text` view. It reads:

```
def model test as
    def unit millisecond as
        unit second {pref: milli};
    enddef;

    def comp component as
        var F: rad_per_sec3;

        F = sel
            case (t > 0{second}) and (t < 1{millisecond}):
                1.0{rad_per_sec3};
            case (t > 1{second}) and (t < 2{second}):
                -1.0{rad_per_sec3};
            otherwise:
                0.0{rad_per_sec3};
        endsel;
    enddef;
enddef;
```

Again, if I validate that file in the `CellML Text` view, everything will be considered fine and, indeed, it is when it comes to the `CellML Text` format. However, if you switch to the `Raw CellML` view, you will a lot of errors.

So, in conclusion (and as mentioned just below the screenshot [here](http://opencor.ws/user/plugins/editing/CellMLTextView.html)):
 - Write your model using the `CellML Text` view, but
 - Validate your model using the `Raw CellML` view.

**Comment by agarny** [comment]
Closing since this is a non-issue.

**Comment by mirams** [comment]
Would be nice to not have to ever use the `Raw CellML` view, and get errors/warnings propagated to the `CellML Text` view... the idea of XML/CellML is that a human never has to read it, but can if they really want to?!

**Comment by agarny** [comment]
@mirams: yes, I find it as frustrating as you do, but as mentioned in the first message: "the CellML validation service cannot be used in [the CellML Text view]." Actually, truth be told, maybe it could, but it would then require a lot of work to get the CellML API to do what we want and need (assuming it could be done!). So, since the CellML API is to be replaced by libCellML, it was agreed that we should wait for the latter to come out to have 'proper' validation in the `CellML Text` view.
